File: 254. Factor Combinations.cpp
class Solution {
public:
    // backtracking: 51.29%, ok
    vector<vector<int>> getFactors(int n) {
        vector<vector<int>> ret;
        dfs(ret, vector<int>(), n);
        return ret;
    }

    void dfs(vector<vector<int>>& ret, vector<int> temp, int n){
        // the smallest factor so far
        int i = temp.empty()? 2: temp.back();
        for(; i * i <= n; i++){
            if(n % i == 0){
                temp.push_back(i);
                dfs(ret, temp, n / i);
                temp.pop_back();
            }
        }
        temp.push_back(n);
        if(temp.size() > 1)
            ret.push_back(temp);
    }
};

File: 254. Factor Combinations.py
import math
class Solution(object):
    
    # dfs, backtracking: O(sqrt(N)*N) time, O(?) space
    def getFactors(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        def dfs(num, factor, curr):
            for i in xrange(factor, int(math.sqrt(num)) + 1):    # bug: right boundry
                if num % i == 0:
                    self.ret.append(curr + [i, num / i])
                    dfs(num / i, i, curr + [i])
        self.ret = []
        dfs(n, 2, [])
        return self.ret
                
        